python/dolfinx/io/gmsh.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (C) 2022 Jørgen S. Dokken
+#
+# This file is part of DOLFINx (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+"""Tools to extract data from GMSH models"""
+import typing
+
+import numpy as np
+
+import ufl
+from dolfinx import cpp as _cpp
+from dolfinx.mesh import CellType
+
+__all__ = ["extract_topology_and_markers", "extract_geometry", "ufl_mesh", "cell_perm"]
+
+
+def extract_topology_and_markers(gmsh_model, model_name=None):
+    """Extract all entities tagged with a physical marker in the gmsh
+    model, and collect the data per cell type. Returns a nested
+    dictionary where the first key is the gmsh MSH element type integer.
+    Each element type present in the model contains the cell topology of
+    the elements and corresponding markers.
+
+    """
+    if model_name is not None:
+        gmsh_model.setCurrent(model_name)
+
+    # Get the physical groups from gmsh on the form [(dim1, tag1),(dim1,
+    # tag2), (dim2, tag3),...]
+    phys_grps = gmsh_model.getPhysicalGroups()
+    topologies = {}
+    for dim, tag in phys_grps:
+        # Get the entities for a given dimension:
+        # dim=0->Points, dim=1->Lines, dim=2->Triangles/Quadrilaterals,
+        # etc.
+        entities = gmsh_model.getEntitiesForPhysicalGroup(dim, tag)
+
+        for entity in entities:
+            # Get data about the elements on a given entity:
+            # NOTE: Assumes that each entity only have one cell-type
+            element_data = gmsh_model.mesh.getElements(dim, tag=entity)
+            element_types, element_tags, node_tags = element_data
+            assert len(element_types) == 1
+
+            # The MSH type of the cells on the element
+            element_type = element_types[0]
+            num_el = len(element_tags[0])
+
+            # Determine number of local nodes per element to create the
+            # topology of the elements
+            properties = gmsh_model.mesh.getElementProperties(element_type)
+            name, dim, order, num_nodes, local_coords, _ = properties
+
+            # 2D array of shape (num_elements,num_nodes_per_element)
+            # containing the topology of the elements on this entity
+            # NOTE: GMSH indexing starts with 1 and not zero
+            element_topology = node_tags[0].reshape(-1, num_nodes) - 1
+
+            # Gather data for each element type and the
```

Comment:
Gather implies communication in the DOLFINx context. Should it be 'Get data on each . . .'